File: packages/capacitor-plugin/ios/Sources/UpdaterPlugin/BundleStore.swift
import Foundation

final class BundleStore {
  /// Release identity is metadata; an installation must never reuse a live directory.
  static func newInstallationId() -> String { "bundle-\(UUID().uuidString.lowercased())" }

  private enum Keys {
    static let currentBundleId = "otakit_current_bundle_id"
    static let fallbackBundleId = "otakit_fallback_bundle_id"
    static let stagedBundleId = "otakit_staged_bundle_id"
    static let lastFailedBundleInfo = "otakit_last_failed_bundle_info"
    static let lastResolvedRuntimeKey = "otakit_last_resolved_runtime_key"
    static let overrideChannel = "otakit_override_channel"
  }

  private let defaults: UserDefaults
  private let rootDirectory: URL
  private let fileManager = FileManager.default
  var appRuntimeVersion: String?

  init(defaults: UserDefaults = .standard, rootDirectory: URL? = nil) {
    self.defaults = defaults
    self.rootDirectory = rootDirectory ?? FileManager.default.urls(
      for: .applicationSupportDirectory,
      in: .userDomainMask
    )[0]
  }

  private var decoder: JSONDecoder {
    let decoder = JSONDecoder()
    decoder.dateDecodingStrategy = .iso8601
    return decoder
  }

  private var encoder: JSONEncoder {
    let encoder = JSONEncoder()
    encoder.dateEncodingStrategy = .iso8601
    return encoder
  }

  private(set) lazy var bundlesDirectory: URL = {
    let directory = rootDirectory.appendingPathComponent(
      "otakit_bundles",
      isDirectory: true
    )
    try? fileManager.createDirectory(
      at: directory,
      withIntermediateDirectories: true
    )
    return directory
  }()

  /// Content-addressed file cache for the deltas strategy (`otakit_files/<sha256>`).
  private(set) lazy var filesCacheDirectory: URL = {
    let directory = rootDirectory.appendingPathComponent(
      "otakit_files",
      isDirectory: true
    )
    try? fileManager.createDirectory(
      at: directory,
      withIntermediateDirectories: true
    )
    return directory
  }()

  var builtinVersion: String {
    Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "0.0.0"
  }

  var nativeBuild: String {
    Bundle.main.infoDictionary?["CFBundleVersion"] as? String ?? "1"
  }

  func builtinBundle() -> BundleInfo {
    BundleInfo(
      id: "builtin",
      version: builtinVersion,
      runtimeVersion: appRuntimeVersion,
      status: .builtin,
      downloadedAt: nil,
      sha256: nil,
      path: nil,
      channel: nil,
      releaseId: nil
    )
  }

  func bundleDirectory(for id: String) -> URL {
    bundlesDirectory.appendingPathComponent(id, isDirectory: true)
  }

  private func metadataURL(for id: String) -> URL {
    bundleDirectory(for: id).appendingPathComponent("bundle.json")
  }

  func bundlePath(id: String) -> String? {
    guard let info = getBundle(id: id) else {
      return nil
    }
    return info.path
  }

  func bundleExists(id: String) -> Bool {
    if id == "builtin" {
      return true
    }
    return getBundle(id: id) != nil
  }

  func saveBundle(_ bundle: BundleInfo) throws {
    guard !bundle.isBuiltin else {
      return
    }

    let directory = bundleDirectory(for: bundle.id)
    try fileManager.createDirectory(at: directory, withIntermediateDirectories: true)
    let data = try encoder.encode(bundle)
    try data.write(to: metadataURL(for: bundle.id), options: .atomic)
  }

  func getBundle(id: String) -> BundleInfo? {
    if id == "builtin" {
      return builtinBundle()
    }

    let metadata = metadataURL(for: id)
    guard let data = try? Data(contentsOf: metadata) else {
      return nil
    }
    return try? decoder.decode(BundleInfo.self, from: data)
  }

  func listDownloadedBundles() -> [BundleInfo] {
    guard
      let ids = try? fileManager.contentsOfDirectory(
        atPath: bundlesDirectory.path
      )
    else {
      return []
    }

    var bundles = ids.compactMap { getBundle(id: $0) }
    bundles.sort { lhs, rhs in
      switch (lhs.downloadedAt, rhs.downloadedAt) {
      case let (left?, right?):
        return left > right
      case (.some, .none):
        return true
      case (.none, .some):
        return false
      case (.none, .none):
        return lhs.id < rhs.id
      }
    }
    return bundles
  }

  private struct CoreState: Codable {
    var version = 1
    var currentId: String?
    var fallbackId: String?
    var stagedId: String?
    var lastFailed: BundleInfo?
  }

  private var stateURL: URL { rootDirectory.appendingPathComponent("otakit-state.json") }
  private let persistenceLock = NSRecursiveLock()

  private func readCoreState() throws -> CoreState {
    persistenceLock.lock()
    defer { persistenceLock.unlock() }
    if fileManager.fileExists(atPath: stateURL.path) {
      let state = try decoder.decode(CoreState.self, from: Data(contentsOf: stateURL))
      guard state.version == 1 else {
        throw CocoaError(.coderReadCorrupt)
      }
      return state
    }
    // Existing installations migrate on their first successful state write.
    let failed = defaults.data(forKey: Keys.lastFailedBundleInfo)
      .flatMap { try? decoder.decode(BundleInfo.self, from: $0) }
    return CoreState(
      currentId: defaults.string(forKey: Keys.currentBundleId),
      fallbackId: defaults.string(forKey: Keys.fallbackBundleId),
      stagedId: defaults.string(forKey: Keys.stagedBundleId),
      lastFailed: failed
    )
  }

  private func writeCoreState(_ state: CoreState) throws {
    try fileManager.createDirectory(at: rootDirectory, withIntermediateDirectories: true)
    try encoder.encode(state).write(to: stateURL, options: .atomic)
  }

  private func mutateCoreState(_ mutation: (inout CoreState) -> Void) throws {
    persistenceLock.lock()
    defer { persistenceLock.unlock() }
    var state = try readCoreState()
    mutation(&state)
    try writeCoreState(state)
  }

  func setCoreState(
    currentId: String?, fallbackId: String?, stagedId: String?, lastFailed: BundleInfo?
  ) throws {
    persistenceLock.lock()
    defer { persistenceLock.unlock() }
    // Validate an existing state before replacing it; corruption must fail closed.
    _ = try readCoreState()
    try writeCoreState(CoreState(
      currentId: currentId, fallbackId: fallbackId, stagedId: stagedId, lastFailed: lastFailed
    ))
  }

  func protectedBundleIds() throws -> Set<String> {
    let state = try readCoreState()
    return Set([state.currentId, state.fallbackId, state.stagedId].compactMap { $0 })
  }

  func getCurrentBundle() -> BundleInfo {
    guard let id = getCurrentBundleId(), let bundle = getBundle(id: id) else {
      return builtinBundle()
    }
    return bundle
  }

  func getCurrentBundleId() -> String? { (try? readCoreState())?.currentId }

  func setCurrentBundleId(_ id: String?) throws {
    try mutateCoreState { $0.currentId = id }
  }

  func getFallbackBundle() -> BundleInfo {
    guard let id = getFallbackBundleId(), let bundle = getBundle(id: id) else {
      return builtinBundle()
    }
    return bundle
  }

  func getFallbackBundleId() -> String? { (try? readCoreState())?.fallbackId }

  func setFallbackBundleId(_ id: String?) throws {
    try mutateCoreState { $0.fallbackId = id }
  }

  func getStagedBundleId() -> String? { (try? readCoreState())?.stagedId }

  func setStagedBundleId(_ id: String?) throws {
    try mutateCoreState { $0.stagedId = id }
  }

  func setLastFailedBundle(_ bundle: BundleInfo?) throws {
    try mutateCoreState { $0.lastFailed = bundle }
  }

  func getLastFailedBundle() -> BundleInfo? { (try? readCoreState())?.lastFailed }

  func getOverrideChannel() -> String? {
    defaults.string(forKey: Keys.overrideChannel)
  }

  func setOverrideChannel(_ channel: String?) {
    if let channel {
      defaults.set(channel, forKey: Keys.overrideChannel)
    } else {
      defaults.removeObject(forKey: Keys.overrideChannel)
    }
  }

  func getLastResolvedRuntimeKey() -> String? {
    defaults.string(forKey: Keys.lastResolvedRuntimeKey)
  }

  func setLastResolvedRuntimeKey(_ runtimeKey: String?) {
    if let runtimeKey {
      defaults.set(runtimeKey, forKey: Keys.lastResolvedRuntimeKey)
    } else {
      // nil clears the key so the next cold start is treated as unresolved again
      defaults.removeObject(forKey: Keys.lastResolvedRuntimeKey)
    }
  }

  func markStatus(bundleId: String, status: BundleStatus) throws {
    guard var bundle = getBundle(id: bundleId) else {
      return
    }
    bundle = BundleInfo(
      id: bundle.id,
      version: bundle.version,
      runtimeVersion: bundle.runtimeVersion,
      status: status,
      downloadedAt: bundle.downloadedAt,
      sha256: bundle.sha256,
      path: bundle.path,
      channel: bundle.channel,
      releaseId: bundle.releaseId
    )
    try saveBundle(bundle)
  }

  func deleteBundle(id: String) throws {
    guard id != "builtin" else {
      return
    }

    try setCoreState(
      currentId: getCurrentBundleId() == id ? nil : getCurrentBundleId(),
      fallbackId: getFallbackBundleId() == id ? nil : getFallbackBundleId(),
      stagedId: getStagedBundleId() == id ? nil : getStagedBundleId(),
      lastFailed: getLastFailedBundle()
    )
    try fileManager.removeItem(at: bundleDirectory(for: id))
  }
}

File: packages/capacitor-plugin/android/src/main/java/com/otakit/updater/BundleStore.java
package com.otakit.updater;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.AtomicFile;
import com.getcapacitor.JSArray;
import java.io.File;
import java.io.FileOutputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import org.json.JSONException;
import org.json.JSONObject;

final class BundleStore {

  private static final String PREFS_NAME = "otakit_updater_state";
  private static final String KEY_CURRENT = "current_bundle_id";
  private static final String KEY_FALLBACK = "fallback_bundle_id";
  private static final String KEY_STAGED = "staged_bundle_id";
  private static final String KEY_LAST_FAILED_BUNDLE_INFO = "last_failed_bundle_info";
  private static final String KEY_LAST_RESOLVED_RUNTIME_KEY = "last_resolved_runtime_key";
  private static final String KEY_OVERRIDE_CHANNEL = "override_channel";

  private final Context context;
  private final SharedPreferences prefs;
  private final File bundlesDirectory;
  private final File filesCacheDirectory;
  private final String builtinVersion;
  private final String nativeBuild;
  private final String appRuntimeVersion;

  BundleStore(
    Context context,
    String builtinVersion,
    String nativeBuild,
    String appRuntimeVersion
  ) {
    this.context = context.getApplicationContext();
    this.prefs = this.context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
    this.builtinVersion = builtinVersion;
    this.nativeBuild = nativeBuild;
    this.appRuntimeVersion = appRuntimeVersion;
    this.bundlesDirectory = new File(this.context.getFilesDir(), "otakit_bundles");
    if (!bundlesDirectory.exists()) {
      //noinspection ResultOfMethodCallIgnored
      bundlesDirectory.mkdirs();
    }
    this.filesCacheDirectory = new File(this.context.getFilesDir(), "otakit_files");
    if (!filesCacheDirectory.exists()) {
      //noinspection ResultOfMethodCallIgnored
      filesCacheDirectory.mkdirs();
    }
  }

  /** Content-addressed file cache for the deltas strategy ({@code otakit_files/<sha256>}). */
  File getFilesCacheDirectory() {
    return filesCacheDirectory;
  }

  String getNativeBuild() {
    return nativeBuild;
  }

  String getBuiltinVersion() {
    return builtinVersion;
  }

  SharedPreferences getPrefs() {
    return prefs;
  }

  BundleInfo builtinBundle() {
    return new BundleInfo(
      "builtin",
      builtinVersion,
      appRuntimeVersion,
      BundleStatus.BUILTIN,
      null,
      null,
      null,
      null,
      null
    );
  }

  File bundleDirectory(String id) {
    return new File(bundlesDirectory, id);
  }

  private File metadataFile(String id) {
    return new File(bundleDirectory(id), "bundle.json");
  }

  synchronized void saveBundle(BundleInfo bundle) throws Exception {
    if (bundle.isBuiltin()) {
      return;
    }

    File directory = bundleDirectory(bundle.id);
    if (!directory.exists() && !directory.mkdirs()) {
      throw new IllegalStateException("Failed to create bundle directory");
    }

    byte[] bytes = bundle.toJSONObject().toString().getBytes(StandardCharsets.UTF_8);
    writeAtomically(metadataFile(bundle.id), bytes);
  }

  synchronized BundleInfo getBundle(String id) {
    if ("builtin".equals(id)) {
      return builtinBundle();
    }

    File metadata = metadataFile(id);
    try {
      byte[] bytes = new AtomicFile(metadata).readFully();
      JSONObject json = new JSONObject(new String(bytes, StandardCharsets.UTF_8));
      return BundleInfo.fromJSONObject(json);
    } catch (Exception ignored) {
      return null;
    }
  }

  synchronized boolean bundleExists(String id) {
    return getBundle(id) != null;
  }

  private File stateFile() {
    return new File(context.getFilesDir(), "otakit-state.json");
  }

  private static void writeAtomically(File file, byte[] bytes) throws Exception {
    AtomicFile atomic = new AtomicFile(file);
    FileOutputStream output = null;
    try {
      output = atomic.startWrite();
      output.write(bytes);
      output.getFD().sync();
      atomic.finishWrite(output);
      output = null;
      if (!Arrays.equals(atomic.readFully(), bytes)) {
        throw new java.io.IOException("Atomic updater state write did not complete");
      }
    } catch (Exception error) {
      if (output != null) atomic.failWrite(output);
      throw error;
    }
  }

  private JSONObject readCoreState() throws Exception {
    File file = stateFile();
    if (file.exists() || new File(file.getPath() + ".bak").exists()) {
      JSONObject state = new JSONObject(
        new String(new AtomicFile(file).readFully(), StandardCharsets.UTF_8)
      );
      if (state.getInt("version") != 1) throw new IllegalStateException(
        "Unsupported updater state"
      );
      return state;
    }
    JSONObject state = new JSONObject();
    state.put("version", 1);
    state.put("currentId", prefs.getString(KEY_CURRENT, null));
    state.put("fallbackId", prefs.getString(KEY_FALLBACK, null));
    state.put("stagedId", prefs.getString(KEY_STAGED, null));
    String failed = prefs.getString(KEY_LAST_FAILED_BUNDLE_INFO, null);
    if (failed != null) {
      try {
        state.put("lastFailed", new JSONObject(failed));
      } catch (JSONException ignored) {}
    }
    return state;
  }

  private JSONObject readCoreStateOrEmpty() {
    try {
      return readCoreState();
    } catch (Exception error) {
      return new JSONObject();
    }
  }

  private void writeCoreState(JSONObject state) throws Exception {
    writeAtomically(stateFile(), state.toString().getBytes(StandardCharsets.UTF_8));
  }

  private void setCoreValue(String key, Object value) {
    try {
      JSONObject state = readCoreState();
      state.put(key, value);
      writeCoreState(state);
    } catch (Exception error) {
      throw new IllegalStateException("Could not persist updater state", error);
    }
  }

  synchronized void setCoreState(
    String currentId,
    String fallbackId,
    String stagedId,
    BundleInfo lastFailed
  ) {
    try {
      readCoreState();
      JSONObject state = new JSONObject();
      state.put("version", 1);
      state.put("currentId", currentId);
      state.put("fallbackId", fallbackId);
      state.put("stagedId", stagedId);
      if (lastFailed != null) state.put("lastFailed", lastFailed.toJSONObject());
      writeCoreState(state);
    } catch (Exception error) {
      throw new IllegalStateException("Could not persist updater state", error);
    }
  }

  synchronized BundleInfo getCurrentBundle() {
    String id = getCurrentBundleId();
    BundleInfo bundle = id == null ? null : getBundle(id);
    return bundle != null ? bundle : builtinBundle();
  }

  synchronized java.util.Set<String> protectedBundleIds() throws Exception {
    JSONObject state = readCoreState();
    java.util.Set<String> ids = new java.util.HashSet<>();
    for (String key : new String[] { "currentId", "fallbackId", "stagedId" }) {
      String id = state.optString(key, null);
      if (id != null) ids.add(id);
    }
    return ids;
  }

  synchronized String getCurrentBundleId() {
    return readCoreStateOrEmpty().optString("currentId", null);
  }

  synchronized void setCurrentBundleId(String id) {
    setCoreValue("currentId", id);
  }

  synchronized BundleInfo getFallbackBundle() {
    String id = getFallbackBundleId();
    BundleInfo bundle = id == null ? null : getBundle(id);
    return bundle != null ? bundle : builtinBundle();
  }

  synchronized String getFallbackBundleId() {
    return readCoreStateOrEmpty().optString("fallbackId", null);
  }

  synchronized void setFallbackBundleId(String id) {
    setCoreValue("fallbackId", id);
  }

  synchronized String getStagedBundleId() {
    return readCoreStateOrEmpty().optString("stagedId", null);
  }

  synchronized void setStagedBundleId(String id) {
    setCoreValue("stagedId", id);
  }

  synchronized void setLastFailedBundle(BundleInfo bundle) {
    try {
      setCoreValue("lastFailed", bundle == null ? null : bundle.toJSONObject());
    } catch (JSONException error) {
      throw new IllegalStateException(error);
    }
  }

  synchronized BundleInfo getLastFailedBundle() {
    JSONObject failed = readCoreStateOrEmpty().optJSONObject("lastFailed");
    return failed == null ? null : BundleInfo.fromJSONObject(failed);
  }

  synchronized String getOverrideChannel() {
    return prefs.getString(KEY_OVERRIDE_CHANNEL, null);
  }

  synchronized void setOverrideChannel(String channel) {
    SharedPreferences.Editor editor = prefs.edit();
    if (channel == null) {
      editor.remove(KEY_OVERRIDE_CHANNEL);
    } else {
      editor.putString(KEY_OVERRIDE_CHANNEL, channel);
    }
    editor.commit();
  }

  synchronized String getLastResolvedRuntimeKey() {
    return prefs.getString(KEY_LAST_RESOLVED_RUNTIME_KEY, null);
  }

  synchronized void setLastResolvedRuntimeKey(String runtimeKey) {
    SharedPreferences.Editor editor = prefs.edit();
    if (runtimeKey == null) {
      // null clears the key so the next cold start is treated as unresolved again
      editor.remove(KEY_LAST_RESOLVED_RUNTIME_KEY);
    } else {
      editor.putString(KEY_LAST_RESOLVED_RUNTIME_KEY, runtimeKey);
    }
    editor.commit();
  }

  synchronized void markStatus(String bundleId, BundleStatus status) {
    BundleInfo existing = getBundle(bundleId);
    if (existing == null) {
      return;
    }
    try {
      saveBundle(existing.withStatus(status));
    } catch (Exception error) {
      throw new IllegalStateException("Could not persist bundle status", error);
    }
  }

  synchronized void deleteBundle(String id) throws Exception {
    if ("builtin".equals(id)) {
      return;
    }

    setCoreState(
      id.equals(getCurrentBundleId()) ? null : getCurrentBundleId(),
      id.equals(getFallbackBundleId()) ? null : getFallbackBundleId(),
      id.equals(getStagedBundleId()) ? null : getStagedBundleId(),
      getLastFailedBundle()
    );
    deleteRecursively(bundleDirectory(id));
  }

  synchronized List<BundleInfo> listDownloadedBundleInfos() {
    List<BundleInfo> result = new ArrayList<>();
    File[] entries = bundlesDirectory.listFiles();
    if (entries != null) {
      Arrays.sort(entries, Comparator.comparing(File::getName));
      for (File entry : entries) {
        BundleInfo bundle = getBundle(entry.getName());
        if (bundle != null) {
          result.add(bundle);
        }
      }
    }
    return result;
  }

  synchronized JSArray listDownloadedBundles() {
    List<BundleInfo> result = listDownloadedBundleInfos();
    JSArray array = new JSArray();
    for (BundleInfo bundle : result) {
      array.put(bundle.toJSObject());
    }
    return array;
  }

  private void deleteRecursively(File target) throws Exception {
    if (!target.exists()) {
      return;
    }
    if (target.isDirectory()) {
      File[] children = target.listFiles();
      if (children != null) {
        for (File child : children) {
          deleteRecursively(child);
        }
      }
    }
    if (!target.delete()) {
      throw new IllegalStateException("Failed to delete " + target.getAbsolutePath());
    }
  }
}
